File: src/main/java/mediathek/gui/dialog/lucene_tutorial/LuceneTutorialDialog.java
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package mediathek.gui.dialog.lucene_tutorial;

import mediathek.config.Konstanten;
import mediathek.gui.actions.UrlHyperlinkAction;
import mediathek.tool.EscapeKeyHandler;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.swing.*;
import javax.swing.event.HyperlinkEvent;
import java.awt.*;
import java.io.InputStream;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;

public final class LuceneTutorialDialog extends JDialog {
    private static final Logger logger = LogManager.getLogger();
    private static final String FALLBACK_HTML = """
            <html><body>
            <p>Die Lucene-Anleitung konnte nicht geladen werden.</p>
            </body></html>
            """;

    public LuceneTutorialDialog(Window owner) {
        super(owner, "Lucene-Suchsyntax", ModalityType.MODELESS);
        initComponents();
        loadTutorial();
        EscapeKeyHandler.installHandler(this, this::dispose);
    }

    private void loadTutorial() {
        try {
            var markdown = loadMarkdownResource();
            if (markdown != null) {
                logger.trace("Rendering Lucene tutorial from Markdown resource {}", Konstanten.PFAD_LUCENE_TUTORIAL_MARKDOWN);
                tutorialPane.setHtml(LuceneTutorialRenderer.renderMarkdown(markdown));
                return;
            }

            logger.error("Lucene tutorial Markdown resource could not be found");
            logger.trace("Showing inline fallback because no Lucene tutorial resource could be loaded");
            showFallbackHtml();
        } catch (Exception ex) {
            logger.error("Failed to load Lucene tutorial resource", ex);
            logger.trace("Showing inline fallback because loading the Lucene tutorial resource failed");
            showFallbackHtml();
        }
    }

    private String loadMarkdownResource() throws Exception {
        try (InputStream markdownStream = getClass().getResourceAsStream(Konstanten.PFAD_LUCENE_TUTORIAL_MARKDOWN)) {
            if (markdownStream == null) {
                return null;
            }
            return new String(markdownStream.readAllBytes(), StandardCharsets.UTF_8);
        }
    }

    private void showFallbackHtml() {
        tutorialPane.setHtml(FALLBACK_HTML);
    }

    private void initComponents() {
        var scrollPane = new JScrollPane();
        tutorialPane = new FlexmarkHtmlPane();

        setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
        setType(Type.UTILITY);
        setMinimumSize(new Dimension(840, 700));

        tutorialPane.addHyperlinkListener(e -> {
            if (e.getEventType() != HyperlinkEvent.EventType.ACTIVATED || e.getURL() == null) {
                return;
            }

            var protocol = e.getURL().getProtocol();
            if ("http".equalsIgnoreCase(protocol) || "https".equalsIgnoreCase(protocol)) {
                try {
                    UrlHyperlinkAction.openURI(e.getURL().toURI());
                } catch (URISyntaxException ex) {
                    logger.error("Failed to open tutorial link {}", e.getURL(), ex);
                }
            }
        });
        scrollPane.setViewportView(tutorialPane);

        var contentPane = getContentPane();
        contentPane.setLayout(new BorderLayout(8, 8));
        contentPane.add(scrollPane, BorderLayout.CENTER);

        pack();
        setLocationRelativeTo(getOwner());
    }

    private FlexmarkHtmlPane tutorialPane;
}

File: src/main/kotlin/mediathek/gui/dialog/lucene_tutorial/LuceneTutorialDialog.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package mediathek.gui.dialog.lucene_tutorial

import mediathek.config.Konstanten
import mediathek.gui.actions.UrlHyperlinkAction
import mediathek.tool.EscapeKeyHandler
import org.apache.logging.log4j.LogManager
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.Window
import java.net.URISyntaxException
import javax.swing.JDialog
import javax.swing.JScrollPane
import javax.swing.WindowConstants
import javax.swing.event.HyperlinkEvent

class LuceneTutorialDialog(
    owner: Window,
    title: String = "Lucene-Suchsyntax",
    private val markdownResourcePath: String = Konstanten.PFAD_LUCENE_TUTORIAL_MARKDOWN,
) : JDialog(owner, title, ModalityType.MODELESS) {
    private val tutorialPane = FlexmarkHtmlPane()

    init {
        initComponents()
        loadTutorial()
        EscapeKeyHandler.installHandler(this) { dispose() }
    }

    private fun loadTutorial() {
        try {
            val markdown = loadMarkdownResource()
            if (markdown != null) {
                logger.trace("Rendering Markdown help from resource {}", markdownResourcePath)
                tutorialPane.setHtml(LuceneTutorialRenderer.renderMarkdown(markdown))
                return
            }

            logger.error("Markdown help resource could not be found: {}", markdownResourcePath)
            logger.trace("Showing inline fallback because no Markdown help resource could be loaded")
            showFallbackHtml()
        } catch (ex: Exception) {
            logger.error("Failed to load Markdown help resource {}", markdownResourcePath, ex)
            logger.trace("Showing inline fallback because loading the Markdown help resource failed")
            showFallbackHtml()
        }
    }

    private fun loadMarkdownResource(): String? =
        javaClass.getResourceAsStream(markdownResourcePath)?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }

    private fun showFallbackHtml() {
        tutorialPane.setHtml(FALLBACK_HTML)
    }

    private fun initComponents() {
        defaultCloseOperation = WindowConstants.DISPOSE_ON_CLOSE
        type = Window.Type.UTILITY
        minimumSize = Dimension(840, 700)

        tutorialPane.addHyperlinkListener { event ->
            if (event.eventType != HyperlinkEvent.EventType.ACTIVATED || event.url == null) {
                return@addHyperlinkListener
            }

            val protocol = event.url.protocol
            if (protocol.equals("http", ignoreCase = true) || protocol.equals("https", ignoreCase = true)) {
                try {
                    UrlHyperlinkAction.openURI(event.url.toURI())
                } catch (ex: URISyntaxException) {
                    logger.error("Failed to open tutorial link {}", event.url, ex)
                }
            }
        }

        contentPane.layout = BorderLayout(8, 8)
        contentPane.add(JScrollPane(tutorialPane), BorderLayout.CENTER)

        pack()
        setLocationRelativeTo(owner)
    }

    private companion object {
        private val logger = LogManager.getLogger()
        private val FALLBACK_HTML = """
            <html><body>
            <p>Die Hilfe konnte nicht geladen werden.</p>
            </body></html>
            """.trimIndent()
    }
}
